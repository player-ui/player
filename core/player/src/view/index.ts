export * from './view';
export * from './resolver';
export * from './parser';
export * from './builder';
export * from './plugins';
