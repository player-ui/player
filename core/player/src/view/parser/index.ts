import { omit, setIn } from 'timm';
import { SyncBailHook, SyncWaterfallHook } from 'tapable-ts';
import type { Template } from '@player-ui/types';
import type { AnyAssetType, Node } from './types';
import { NodeType } from './types';
import { getNodeID, hasAsync } from './utils';

export * from './types';
export * from './utils';

export const EMPTY_NODE: Node.Empty = {
  type: NodeType.Empty,
};

export interface ParseObjectOptions {
  /** how nested the templated is */
  templateDepth?: number;
}

interface NestedObj {
  /** The values of a nested local object */
  children: Node.Child[];

  value: any;
}
/**
 * The Parser is the way to take an incoming view from the user and parse it into an AST.
 * It provides a few ways to interact with the parsing, including mutating an object before and after creation of an AST node
 */
export class Parser {
  public readonly hooks = {
    /**
     * A hook to interact with an object _before_ parsing it into an AST
     *
     * @param value - The object we're are about to parse
     * @returns - A new value to parse.
     *  If undefined, the original value is used.
     *  If null, we stop parsing this node.
     */
    onParseObject: new SyncWaterfallHook<[object, NodeType]>(),

    /**
     * A callback to interact with an AST _after_ we parse it into the AST
     *
     * @param value - The object we parsed
     * @param node - The AST node we generated
     * @returns - A new AST node to use
     *   If undefined, the original value is used.
     *   If null, we ignore this node all together
     */
    onCreateASTNode: new SyncWaterfallHook<
      [Node.Node | undefined | null, object]
    >(),

    determineNodeType: new SyncBailHook<[object | string], NodeType>(),

    parseNode: new SyncBailHook<
      [
        obj: object,
        nodeType: Node.ChildrenTypes,
        parseOptions: ParseObjectOptions,
        determinedNodeType: NodeType | null
      ],
      Node.Node
    >(),
  };

  public parseView(value: AnyAssetType): Node.View {
    const viewNode = this.parseObject(value, NodeType.View);

    if (!viewNode) {
      throw new Error('Unable to parse object into a view');
    }

    return viewNode as Node.View;
  }

  private parseAsync(
    obj: object,
    type: Node.ChildrenTypes,
    options: ParseObjectOptions
  ): Node.Node | null {
    const parsedAsync = this.parseObject(omit(obj, 'async'), type, options);
    const parsedNodeId = getNodeID(parsedAsync);
    if (parsedAsync !== null && parsedNodeId) {
      return this.createASTNode(
        {
          id: parsedNodeId,
          type: NodeType.Async,
          value: parsedAsync,
        },
        obj
      );
    }

    return null;
  }

  public createASTNode(node: Node.Node | null, value: any): Node.Node | null {
    const tapped = this.hooks.onCreateASTNode.call(node, value);

    if (tapped === undefined) {
      return node;
    }

    return tapped;
  }

  /**
   * Checks if there are templated values in the object
   *
   * @param obj - The Parsed Object to check to see if we have a template array type for
   * @param localKey - The key being checked
   */
  private hasTemplateValues(obj: any, localKey: string) {
    return (
      Object.hasOwnProperty.call(obj, 'template') &&
      Array.isArray(obj?.template) &&
      obj.template.length &&
      obj.template.find((tmpl: any) => tmpl.output === localKey)
    );
  }

  public parseObject(
    obj: object,
    type: Node.ChildrenTypes = NodeType.Value,
    options: ParseObjectOptions = { templateDepth: 0 }
  ): Node.Node | null {
    const nodeType = this.hooks.determineNodeType.call(obj);

    if (nodeType !== undefined) {
      const parsedNode = this.hooks.parseNode.call(
        obj,
        type,
        options,
        nodeType
      );
      if (parsedNode) {
        return parsedNode;
      }
    }

    const parseLocalObject = (
      currentValue: any,
      objToParse: unknown,
      path: string[] = []
    ): NestedObj => {
      if (typeof objToParse !== 'object' || objToParse === null) {
        return { value: objToParse, children: [] };
      }

      const localObj = this.hooks.onParseObject.call(objToParse, type);

      if (!localObj) {
        return currentValue;
      }

      const objEntries = Array.isArray(localObj)
        ? localObj.map((v, i) => [i, v])
        : [
            ...Object.entries(localObj),
            ...Object.getOwnPropertySymbols(localObj).map((s) => [
              s,
              (localObj as any)[s],
            ]),
          ];

      const defaultValue: NestedObj = {
        children: [],
        value: currentValue,
      };

      const newValue = objEntries.reduce((accumulation, current): NestedObj => {
        const { children, ...rest } = accumulation;
        const [localKey, localValue] = current;
        if (localKey === 'asset' && typeof localValue === 'object') {
          const assetAST = this.parseObject(
            localValue,
            NodeType.Asset,
            options
          );

          if (assetAST) {
            return {
              ...rest,
              children: [
                ...children,
                {
                  path: [...path, 'asset'],
                  value: assetAST,
                },
              ],
            };
          }
        } else if (
          this.hooks.determineNodeType.call(localKey) === NodeType.Template &&
          Array.isArray(localValue)
        ) {
          const templateChildren = localValue
            .map((template: Template) => {
              const templateAST = this.hooks.onCreateASTNode.call(
                {
                  type: NodeType.Template,
                  depth: options.templateDepth ?? 0,
                  data: template.data,
                  template: template.value,
                  dynamic: template.dynamic ?? false,
                },
                template
              );

              if (templateAST?.type === NodeType.MultiNode) {
                templateAST.values.forEach((v) => {
                  // eslint-disable-next-line no-param-reassign
                  v.parent = templateAST;
                });
              }

              if (templateAST) {
                return {
                  path: [...path, template.output],
                  value: templateAST,
                };
              }

              // eslint-disable-next-line no-useless-return
              return;
            })
            .filter((element) => !!element);

          return {
            ...rest,
            children: [...children, ...templateChildren],
          } as NestedObj;
        } else if (
          localValue &&
          this.hooks.determineNodeType.call(localValue) === NodeType.Switch
        ) {
          const localSwitch = this.hooks.parseNode.call(
            localValue,
            NodeType.Value,
            options,
            NodeType.Switch
          );

          if (
            localSwitch &&
            localSwitch.type === NodeType.Value &&
            localSwitch.children?.length === 1 &&
            localSwitch.value === undefined
          ) {
            const firstChild = localSwitch.children[0];

            return {
              ...rest,
              children: [
                ...children,
                {
                  path: [...path, localKey, ...firstChild.path],
                  value: firstChild.value,
                },
              ],
            };
          }

          if (localSwitch) {
            return {
              ...rest,
              children: [
                ...children,
                {
                  path: [...path, localKey],
                  value: localSwitch,
                },
              ],
            };
          }
        } else if (localValue && hasAsync(localValue)) {
          const localAsync = this.parseAsync(
            localValue,
            NodeType.Value,
            options
          );
          if (localAsync) {
            children.push({
              path: [...path, localKey],
              value: localAsync,
            });
          }
        } else if (localValue && Array.isArray(localValue)) {
          const childValues = localValue
            .map((childVal) =>
              this.parseObject(childVal, NodeType.Value, options)
            )
            .filter((child): child is Node.Node => !!child);

          if (childValues.length > 0) {
            const multiNode = this.hooks.onCreateASTNode.call(
              {
                type: NodeType.MultiNode,
                override: !this.hasTemplateValues(localObj, localKey),
                values: childValues,
              },
              localValue
            );

            if (multiNode?.type === NodeType.MultiNode) {
              multiNode.values.forEach((v) => {
                // eslint-disable-next-line no-param-reassign
                v.parent = multiNode;
              });
            }

            if (multiNode) {
              return {
                ...rest,
                children: [
                  ...children,
                  {
                    path: [...path, localKey],
                    value: multiNode,
                  },
                ],
              };
            }
          }
        } else if (localValue && typeof localValue === 'object') {
          const determineNodeType =
            this.hooks.determineNodeType.call(localValue);

          if (determineNodeType === NodeType.Applicability) {
            const parsedNode = this.hooks.parseNode.call(
              localValue,
              NodeType.Value,
              options,
              determineNodeType
            );
            if (parsedNode) {
              return {
                ...rest,
                children: [
                  ...children,
                  {
                    path: [...path, localKey],
                    value: parsedNode,
                  },
                ],
              };
            }
          } else {
            const result = parseLocalObject(accumulation.value, localValue, [
              ...path,
              localKey,
            ]);
            return {
              value: result.value,
              children: [...children, ...result.children],
            };
          }
        } else {
          const value = setIn(
            accumulation.value,
            [...path, localKey],
            localValue
          );

          return {
            children,
            value,
          };
        }

        return accumulation;
      }, defaultValue);

      return newValue;
    };

    const { value, children } = parseLocalObject(undefined, obj);

    const baseAst =
      value === undefined && children.length === 0
        ? undefined
        : {
            type,
            value,
          };

    if (baseAst !== undefined && children.length > 0) {
      const parent = baseAst as Node.BaseWithChildren<any>;
      parent.children = children;
      children.forEach((child) => {
        // eslint-disable-next-line no-param-reassign
        child.value.parent = parent;
      });
    }

    return this.hooks.onCreateASTNode.call(baseAst, obj) ?? null;
  }
}
