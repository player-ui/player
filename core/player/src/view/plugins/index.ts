export { default as TemplatePlugin } from "./template";
export { default as StringResolverPlugin } from "./string-resolver";
export { default as ApplicabilityPlugin } from "./applicability";
export { default as SwitchPlugin } from "./switch";
export { default as MultiNodePlugin } from "./multi-node";
export { default as AssetPlugin } from "./asset";
