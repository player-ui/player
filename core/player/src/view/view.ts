import { SyncHook } from 'tapable-ts';
import type { View as ViewType } from '@player-ui/types';
import type { BindingInstance, BindingFactory } from '../binding';
import type { ValidationProvider, ValidationObject } from '../validator';
import type { Logger } from '../logger';
import type { Resolve } from './resolver';
import { Resolver, toNodeResolveOptions } from './resolver';
import type { Node } from './parser';
import { Parser } from './parser';
import {
  TemplatePlugin,
  StringResolverPlugin,
  ApplicabilityPlugin,
  SwitchPlugin,
} from './plugins';

/**
 * Manages the view level validations
 */
class CrossfieldProvider implements ValidationProvider {
  private allValidations = new Set<ValidationObject>();
  private byBinding = new Map<BindingInstance, Array<ValidationObject>>();
  private logger?: Logger;

  constructor(initialView: ViewType, parser: BindingFactory, logger?: Logger) {
    this.logger = logger;
    this.parse(initialView, parser);
  }

  private parse(contentView: ViewType, parser: BindingFactory) {
    const xfieldRefs = contentView.validation;

    if (xfieldRefs === undefined) {
      return;
    }

    if (!Array.isArray(xfieldRefs)) {
      this.logger?.warn(
        `Unable to register view validations for id: ${contentView.id}. 'validation' property must be an Array.`
      );

      return;
    }

    // Grab the validations from the view (as authored) and parse out the ones that have a _ref_ (to a binding)
    // Group them all by binding to make it easier to return than later

    xfieldRefs.forEach((vRef) => {
      // x-field validations by default are triggered by navigating away from the page
      // the reference can also override that _or_ the severity
      const withDefaults: ValidationObject = {
        trigger: 'navigation',
        severity: 'error',
        ...vRef,
      };

      this.allValidations.add(withDefaults);

      // The validation reference contains a _ref_ (a binding)
      const { ref } = vRef;

      if (ref) {
        /** Group together validations by binding */
        const parsed = parser(ref);

        if (this.byBinding.has(parsed)) {
          this.byBinding.get(parsed)?.push(withDefaults);
        } else {
          this.byBinding.set(parsed, [withDefaults]);
        }
      }
    });
  }

  getValidationsForBinding(binding: BindingInstance) {
    return this.byBinding.get(binding);
  }
}

/** A stateful view instance from an content */
export class ViewInstance implements ValidationProvider {
  public hooks = {
    onUpdate: new SyncHook<[ViewType]>(),
    parser: new SyncHook<[Parser]>(),
    resolver: new SyncHook<[Resolver]>(),
    templatePlugin: new SyncHook<[TemplatePlugin]>(),
  };

  private resolver?: Resolver;
  public readonly initialView: ViewType;
  public readonly resolverOptions: Resolve.ResolverOptions;
  private rootNode?: Node.Node;

  private validationProvider?: CrossfieldProvider;

  private templatePlugin: TemplatePlugin;

  // TODO might want to add a version/timestamp to this to compare updates
  public lastUpdate: Record<string, any> | undefined;

  constructor(initialView: ViewType, resolverOptions: Resolve.ResolverOptions) {
    this.initialView = initialView;
    this.resolverOptions = resolverOptions;
    const pluginOptions = toNodeResolveOptions(resolverOptions);
    new SwitchPlugin(pluginOptions).apply(this);
    new ApplicabilityPlugin().apply(this);
    new StringResolverPlugin().apply(this);
    this.templatePlugin = new TemplatePlugin(pluginOptions);
    this.templatePlugin.apply(this);
  }

  public updateAsync() {
    const update = this.resolver?.update();
    this.lastUpdate = update;
    this.hooks.onUpdate.call(update);
  }

  public update(changes?: Set<BindingInstance>) {
    if (this.rootNode === undefined) {
      /** On initialization of the view, also create a validation parser */
      this.validationProvider = new CrossfieldProvider(
        this.initialView,
        this.resolverOptions.parseBinding,
        this.resolverOptions.logger
      );

      this.hooks.templatePlugin.call(this.templatePlugin);

      const parser = new Parser();
      this.hooks.parser.call(parser);
      this.rootNode = parser.parseView(this.initialView);

      this.resolver = new Resolver(this.rootNode, {
        ...this.resolverOptions,
        parseNode: parser.parseObject.bind(parser),
      });
      this.hooks.resolver.call(this.resolver);
    }

    const update = this.resolver?.update(changes);

    if (this.lastUpdate === update) {
      return this.lastUpdate;
    }

    this.lastUpdate = update;
    this.hooks.onUpdate.call(update);

    return update;
  }

  getValidationsForBinding(binding: BindingInstance) {
    return this.validationProvider?.getValidationsForBinding(binding);
  }
}
