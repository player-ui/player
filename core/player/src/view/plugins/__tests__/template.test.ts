import { describe, it, expect, beforeEach } from "vitest";
import { BindingParser } from "../../../binding";
import type { DataModelWithParser } from "../../../data";
import { LocalModel, withParser } from "../../../data";
import { ExpressionEvaluator } from "../../../expressions";
import { SchemaController } from "../../../schema";
import { NodeType } from "../../parser";
import { Parser } from "../../parser";
import { ViewInstance } from "../../view";
import type { Options } from "../options";
import TemplatePlugin from "../template-plugin";
import { StringResolverPlugin, toNodeResolveOptions } from "../..";

const templateJoinValues = {
  id: "snippet-of-json",
  topic: "Snippet",
  schema: {},
  data: {
    forms: {
      "1099-A": [
        {
          description: "Desciption of concept 1099 1",
          amount: "Help",
        },
      ],
      "1099-B": [
        {
          description: "Desciption of concept 1099 2",
          amount: "Help",
        },
      ],
    },
  },
  views: [
    {
      id: "overviewGroup",
      type: "overviewGroup",
      metaData: {
        role: "stateful",
      },
      modifiers: [
        {
          type: "tag",
          value: "fancy-header",
        },
      ],
      headers: {
        label: {
          asset: {
            id: "line-of-work-summary-gh-header-label",
            type: "text",
            value: "Header",
          },
        },
        values: [
          {
            asset: {
              id: "line-of-work-summary-gh-expenses-simple-header-previous-year",
              type: "text",
              value: "Type",
            },
          },
          {
            asset: {
              id: "line-of-work-summary-gh-expenses-simple-header-cy",
              type: "text",
              value: "2022",
            },
          },
        ],
      },
      template: [
        {
          data: "forms.1099-A",
          output: "values",
          value: {
            asset: {
              id: "overviewItem3",
              type: "overviewItem",
              label: {
                asset: {
                  id: "overviewItem3-label",
                  type: "text",
                  value: "1099-A",
                },
              },
              values: [
                {
                  asset: {
                    id: "overviewItem3-year",
                    type: "text",
                    value: "Desciption of concept 1099 1",
                  },
                },
                {
                  asset: {
                    id: "loverviewItem3-cy",
                    type: "text",
                    value: "4000",
                  },
                },
              ],
            },
          },
        },
        {
          data: "forms.1099-B",
          output: "values",
          value: {
            asset: {
              id: "overviewItem4",
              type: "overviewItem",
              label: {
                asset: {
                  id: "overviewItem4-label",
                  type: "text",
                  value: "1099-B",
                },
              },
              values: [
                {
                  asset: {
                    id: "overviewItem4-year",
                    type: "text",
                    value: "Desciption of concept 1099 2",
                  },
                },
                {
                  asset: {
                    id: "loverviewItem3-cy",
                    type: "text",
                    value: "6000",
                  },
                },
              ],
            },
          },
        },
      ],
      values: [
        {
          asset: {
            id: "overviewItem1",
            type: "overviewItem",
            label: {
              asset: {
                id: "overviewItem1-label",
                type: "text",
                value: "First Summary",
              },
            },
            values: [
              {
                asset: {
                  id: "overviewItem1-year",
                  type: "text",
                  value: "Desciption of year summary 1",
                },
              },
              {
                asset: {
                  id: "loverviewItem1-cy",
                  type: "text",
                  value: "14000",
                },
              },
            ],
          },
        },
        {
          asset: {
            id: "overviewItem2",
            type: "overviewItem",
            label: {
              asset: {
                id: "overviewItem2-label",
                type: "text",
                value: "Second year Summary",
              },
            },
            values: [
              {
                asset: {
                  id: "overviewItem2-year",
                  type: "text",
                  value: "Desciption of year summary item 2",
                },
              },
              {
                asset: {
                  id: "loverviewItem1-cy",
                  type: "text",
                  value: "19000",
                },
              },
            ],
          },
        },
      ],
    },
    {
      id: "overviewGroup",
      type: "overviewGroup",
      metaData: {
        role: "stateful",
      },
      modifiers: [
        {
          type: "tag",
          value: "fancy-header",
        },
      ],
      headers: {
        label: {
          asset: {
            id: "line-of-work-summary-gh-header-label",
            type: "text",
            value: "Header",
          },
        },
        values: [
          {
            asset: {
              id: "line-of-work-summary-gh-expenses-simple-header-previous-year",
              type: "text",
              value: "Type",
            },
          },
          {
            asset: {
              id: "line-of-work-summary-gh-expenses-simple-header-cy",
              type: "text",
              value: "2022",
            },
          },
        ],
      },
      values: [
        {
          asset: {
            id: "overviewItem1",
            type: "overviewItem",
            label: {
              asset: {
                id: "overviewItem1-label",
                type: "text",
                value: "First Summary",
              },
            },
            values: [
              {
                asset: {
                  id: "overviewItem1-year",
                  type: "text",
                  value: "Desciption of year summary 1",
                },
              },
              {
                asset: {
                  id: "loverviewItem1-cy",
                  type: "text",
                  value: "14000",
                },
              },
            ],
          },
        },
        {
          asset: {
            id: "overviewItem2",
            type: "overviewItem",
            label: {
              asset: {
                id: "overviewItem2-label",
                type: "text",
                value: "Second year Summary",
              },
            },
            values: [
              {
                asset: {
                  id: "overviewItem2-year",
                  type: "text",
                  value: "Desciption of year summary item 2",
                },
              },
              {
                asset: {
                  id: "loverviewItem1-cy",
                  type: "text",
                  value: "19000",
                },
              },
            ],
          },
        },
      ],
      template: [
        {
          data: "forms.1099-A",
          output: "values",
          value: {
            asset: {
              id: "overviewItem3",
              type: "overviewItem",
              label: {
                asset: {
                  id: "overviewItem3-label",
                  type: "text",
                  value: "1099-A",
                },
              },
              values: [
                {
                  asset: {
                    id: "overviewItem3-year",
                    type: "text",
                    value: "Desciption of concept 1099 1",
                  },
                },
                {
                  asset: {
                    id: "loverviewItem3-cy",
                    type: "text",
                    value: "4000",
                  },
                },
              ],
            },
          },
        },
        {
          data: "forms.1099-B",
          output: "values",
          value: {
            asset: {
              id: "overviewItem4",
              type: "overviewItem",
              label: {
                asset: {
                  id: "overviewItem4-label",
                  type: "text",
                  value: "1099-B",
                },
              },
              values: [
                {
                  asset: {
                    id: "overviewItem4-year",
                    type: "text",
                    value: "Desciption of concept 1099 2",
                  },
                },
                {
                  asset: {
                    id: "loverviewItem3-cy",
                    type: "text",
                    value: "6000",
                  },
                },
              ],
            },
          },
        },
      ],
    },
  ],
  navigation: {
    BEGIN: "SnippetFlow",
    SnippetFlow: {
      startState: "VIEW_Snippet-View1",
      "VIEW_Snippet-View1": {
        ref: "overviewGroup",
        state_type: "VIEW",
      },
    },
  },
};

const parseBinding = new BindingParser().parse;

describe("templates", () => {
  let model: DataModelWithParser;
  let expressionEvaluator: ExpressionEvaluator;
  let options: Options;
  let parser: Parser;

  beforeEach(() => {
    model = withParser(new LocalModel(), parseBinding);
    expressionEvaluator = new ExpressionEvaluator({
      model,
    });
    parser = new Parser();
    options = {
      evaluate: expressionEvaluator.evaluate,
      schema: new SchemaController(),
      data: {
        format: (binding, val) => val,
        formatValue: (val) => val,
        model,
      },
    };
    new TemplatePlugin(options).applyParser(parser);
  });

  it("works with simple ones", () => {
    const petNames = ["Ginger", "Daisy", "Afra"];
    model.set([["foo.bar", petNames]]);

    expect(
      parser.parseObject({
        id: "foo",
        type: "collection",
        template: [
          {
            data: "foo.bar",
            output: "values",
            value: {
              value: "{{foo.bar._index_}}",
            },
          },
        ],
      }),
    ).toMatchSnapshot();
  });

  it("works with nested templates", () => {
    const petNames = ["Ginger", "Daisy", "Afra"];
    model.set([["foo.pets", petNames]]);

    const peopleNames = ["Adam", "Jenny"];
    model.set([["foo.people", peopleNames]]);

    expect(
      parser.parseObject({
        id: "foo",
        type: "collection",
        template: [
          {
            data: "foo.pets",
            output: "values",
            value: {
              asset: {
                type: "collection",
                id: "outer-collection-_index_",
                template: [
                  {
                    data: "foo.people",
                    output: "values",
                    value: {
                      text: "{{foo.pets._index_}} + {{foo.people._index1_}}",
                    },
                  },
                ],
              },
            },
          },
        ],
      }),
    ).toMatchSnapshot();
  });

  it("determines if nodeType is template", () => {
    const nodeTest = "template";
    const nodeType = parser.hooks.determineNodeType.call(nodeTest);
    expect(nodeType).toStrictEqual("template");
  });

  it("Does not return a nodeType", () => {
    const nodeTest = {
      value: "foo",
    };
    const nodeType = parser.hooks.determineNodeType.call(nodeTest);
    expect(nodeType).toBe(undefined);
  });

  it("returns templateNode if template exists", () => {
    const obj = {
      dynamic: true,
      data: "foo.bar",
      output: "values",
      value: {
        value: "{{foo.bar._index_}}",
      },
    };
    const nodeOptions = {
      templateDepth: 1,
    };
    const parsedNode = parser.hooks.parseNode.call(
      obj,
      NodeType.Value,
      nodeOptions,
      NodeType.Template,
    );
    expect(parsedNode).toStrictEqual({
      data: "foo.bar",
      depth: 1,
      dynamic: true,
      template: {
        value: "{{foo.bar._index_}}",
      },
      type: "template",
    });
  });

  it("returns templateNode if template exists, and templateDepth is not set", () => {
    const obj = {
      data: "foo.bar2",
      output: "values",
      dynamic: true,
      value: {
        value: "{{foo.bar2._index_}}",
      },
    };
    const nodeOptions = {};
    const parsedNode = parser.hooks.parseNode.call(
      obj,
      NodeType.Value,
      nodeOptions,
      NodeType.Template,
    );
    expect(parsedNode).toStrictEqual({
      data: "foo.bar2",
      depth: 0,
      dynamic: true,
      template: {
        value: "{{foo.bar2._index_}}",
      },
      type: "template",
    });
  });
});

describe("dynamic templates", () => {
  it("static - nodes are not updated", () => {
    const petNames = ["Ginger", "Vokey"];
    const model = withParser(new LocalModel({}), parseBinding);
    const evaluator = new ExpressionEvaluator({ model });
    const schema = new SchemaController();

    const view = new ViewInstance(
      {
        id: "my-view",
        asset: {
          id: "foo",
          type: "collection",
          template: [
            {
              dynamic: false,
              data: "foo.bar",
              output: "values",
              value: {
                value: "{{foo.bar._index_}}",
              },
            },
          ],
        },
      } as any,
      {
        model,
        parseBinding,
        evaluator,
        schema,
      },
    );

    const pluginOptions = toNodeResolveOptions(view.resolverOptions);
    new TemplatePlugin(pluginOptions).apply(view);
    new StringResolverPlugin().apply(view);

    model.set([["foo.bar", petNames]]);

    const resolved = view.update();

    expect(resolved).toStrictEqual({
      id: "my-view",
      asset: {
        id: "foo",
        type: "collection",
        values: ["Ginger", "Vokey"].map((value) => ({ value })),
      },
    });

    model.set([["foo.bar", ["Ginger", "Vokey", "Harry"]]]);

    let updated = view.update();
    expect(updated).toStrictEqual({
      id: "my-view",
      asset: {
        id: "foo",
        type: "collection",
        values: ["Ginger", "Vokey"].map((value) => ({ value })),
      },
    });

    model.set([["foo.bar", ["Ginger"]]]);
    updated = view.update();
    expect(updated).toStrictEqual({
      id: "my-view",
      asset: {
        id: "foo",
        type: "collection",
        values: ["Ginger", undefined].map((value) => ({ value })),
      },
    });
  });

  it("dynamic - nodes are updated", () => {
    const petNames = ["Ginger", "Vokey"];
    const model = withParser(new LocalModel({}), parseBinding);
    const evaluator = new ExpressionEvaluator({ model });
    const schema = new SchemaController();

    const view = new ViewInstance(
      {
        id: "my-view",
        asset: {
          id: "foo",
          type: "collection",
          template: [
            {
              dynamic: true,
              data: "foo.bar",
              output: "values",
              value: {
                value: "{{foo.bar._index_}}",
              },
            },
          ],
        },
      } as any,
      {
        model,
        parseBinding,
        evaluator,
        schema,
      },
    );

    const pluginOptions = toNodeResolveOptions(view.resolverOptions);
    new TemplatePlugin(pluginOptions).apply(view);
    new StringResolverPlugin().apply(view);

    model.set([["foo.bar", petNames]]);

    const resolved = view.update();

    expect(resolved).toStrictEqual({
      id: "my-view",
      asset: {
        id: "foo",
        type: "collection",
        values: ["Ginger", "Vokey"].map((value) => ({ value })),
      },
    });

    const barBinding = parseBinding("foo.bar");
    model.set([[barBinding, ["Vokey", "Louis", "Bob"]]]);

    let updated = view.update();
    expect(updated).toStrictEqual({
      id: "my-view",
      asset: {
        id: "foo",
        type: "collection",
        values: ["Vokey", "Louis", "Bob"].map((value) => ({ value })),
      },
    });

    model.set([[barBinding, ["Nuri"]]]);
    updated = view.update();
    expect(updated).toStrictEqual({
      id: "my-view",
      asset: {
        id: "foo",
        type: "collection",
        values: ["Nuri"].map((value) => ({ value })),
      },
    });
  });

  describe("Works with template items plus value items", () => {
    const model = withParser(
      new LocalModel(templateJoinValues.data),
      parseBinding,
    );
    const evaluator = new ExpressionEvaluator({ model });

    it("Should show template item first when coming before values on lexical order", () => {
      const view = new ViewInstance(templateJoinValues.views[0], {
        model,
        parseBinding,
        evaluator,
        schema: new SchemaController(),
      });

      const pluginOptions = toNodeResolveOptions(view.resolverOptions);
      new TemplatePlugin(pluginOptions).apply(view);
      new StringResolverPlugin().apply(view);

      const resolved = view.update();

      expect(resolved.values).toHaveLength(4);
      expect(resolved.values[0]).toMatchSnapshot();
    });
    it("Should show template item last when coming after values on lexical order", () => {
      const view = new ViewInstance(templateJoinValues.views[1], {
        model,
        parseBinding,
        evaluator,
        schema: new SchemaController(),
      });

      const pluginOptions = toNodeResolveOptions(view.resolverOptions);
      new TemplatePlugin(pluginOptions).apply(view);
      new StringResolverPlugin().apply(view);

      const resolved = view.update();

      expect(resolved.values).toHaveLength(4);
      expect(resolved.values[0]).toMatchSnapshot();
    });
  });
});
