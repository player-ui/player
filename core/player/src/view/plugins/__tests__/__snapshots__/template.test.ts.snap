// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dynamic templates Works with template items plus value items Should show template item first when coming before values on lexical order 1`] = `
Object {
  "asset": Object {
    "id": "overviewItem3",
    "label": Object {
      "asset": Object {
        "id": "overviewItem3-label",
        "type": "text",
        "value": "1099-A",
      },
    },
    "type": "overviewItem",
    "values": Array [
      Object {
        "asset": Object {
          "id": "overviewItem3-year",
          "type": "text",
          "value": "Desciption of concept 1099 1",
        },
      },
      Object {
        "asset": Object {
          "id": "loverviewItem3-cy",
          "type": "text",
          "value": "4000",
        },
      },
    ],
  },
}
`;

exports[`dynamic templates Works with template items plus value items Should show template item last when coming after values on lexical order 1`] = `
Object {
  "asset": Object {
    "id": "overviewItem1",
    "label": Object {
      "asset": Object {
        "id": "overviewItem1-label",
        "type": "text",
        "value": "First Summary",
      },
    },
    "type": "overviewItem",
    "values": Array [
      Object {
        "asset": Object {
          "id": "overviewItem1-year",
          "type": "text",
          "value": "Desciption of year summary 1",
        },
      },
      Object {
        "asset": Object {
          "id": "loverviewItem1-cy",
          "type": "text",
          "value": "14000",
        },
      },
    ],
  },
}
`;

exports[`templates works with nested templates 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "values",
      ],
      "value": Object {
        "override": false,
        "parent": [Circular],
        "type": "multi-node",
        "values": Array [
          Object {
            "children": Array [
              Object {
                "path": Array [
                  "asset",
                ],
                "value": Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "values",
                      ],
                      "value": Object {
                        "override": false,
                        "parent": [Circular],
                        "type": "multi-node",
                        "values": Array [
                          Object {
                            "parent": [Circular],
                            "type": "value",
                            "value": Object {
                              "text": "{{foo.pets.0}} + {{foo.people.0}}",
                            },
                          },
                          Object {
                            "parent": [Circular],
                            "type": "value",
                            "value": Object {
                              "text": "{{foo.pets.0}} + {{foo.people.1}}",
                            },
                          },
                        ],
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "asset",
                  "value": Object {
                    "id": "outer-collection-0",
                    "type": "collection",
                  },
                },
              },
            ],
            "parent": [Circular],
            "type": "value",
            "value": undefined,
          },
          Object {
            "children": Array [
              Object {
                "path": Array [
                  "asset",
                ],
                "value": Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "values",
                      ],
                      "value": Object {
                        "override": false,
                        "parent": [Circular],
                        "type": "multi-node",
                        "values": Array [
                          Object {
                            "parent": [Circular],
                            "type": "value",
                            "value": Object {
                              "text": "{{foo.pets.1}} + {{foo.people.0}}",
                            },
                          },
                          Object {
                            "parent": [Circular],
                            "type": "value",
                            "value": Object {
                              "text": "{{foo.pets.1}} + {{foo.people.1}}",
                            },
                          },
                        ],
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "asset",
                  "value": Object {
                    "id": "outer-collection-1",
                    "type": "collection",
                  },
                },
              },
            ],
            "parent": [Circular],
            "type": "value",
            "value": undefined,
          },
          Object {
            "children": Array [
              Object {
                "path": Array [
                  "asset",
                ],
                "value": Object {
                  "children": Array [
                    Object {
                      "path": Array [
                        "values",
                      ],
                      "value": Object {
                        "override": false,
                        "parent": [Circular],
                        "type": "multi-node",
                        "values": Array [
                          Object {
                            "parent": [Circular],
                            "type": "value",
                            "value": Object {
                              "text": "{{foo.pets.2}} + {{foo.people.0}}",
                            },
                          },
                          Object {
                            "parent": [Circular],
                            "type": "value",
                            "value": Object {
                              "text": "{{foo.pets.2}} + {{foo.people.1}}",
                            },
                          },
                        ],
                      },
                    },
                  ],
                  "parent": [Circular],
                  "type": "asset",
                  "value": Object {
                    "id": "outer-collection-2",
                    "type": "collection",
                  },
                },
              },
            ],
            "parent": [Circular],
            "type": "value",
            "value": undefined,
          },
        ],
      },
    },
  ],
  "type": "value",
  "value": Object {
    "id": "foo",
    "type": "collection",
  },
}
`;

exports[`templates works with simple ones 1`] = `
Object {
  "children": Array [
    Object {
      "path": Array [
        "values",
      ],
      "value": Object {
        "override": false,
        "parent": [Circular],
        "type": "multi-node",
        "values": Array [
          Object {
            "parent": [Circular],
            "type": "value",
            "value": Object {
              "value": "{{foo.bar.0}}",
            },
          },
          Object {
            "parent": [Circular],
            "type": "value",
            "value": Object {
              "value": "{{foo.bar.1}}",
            },
          },
          Object {
            "parent": [Circular],
            "type": "value",
            "value": Object {
              "value": "{{foo.bar.2}}",
            },
          },
        ],
      },
    },
  ],
  "type": "value",
  "value": Object {
    "id": "foo",
    "type": "collection",
  },
}
`;
